core/src/main/java/org/apache/iceberg/MetadataUpdateParser.java

```diff
@@ -0,0 +1,163 @@
+      .put(MetadataUpdate.RemoveSnapshotRef.class, REMOVE_SNAPSHOT_REF)
+      .put(MetadataUpdate.SetSnapshotRef.class, SET_SNAPSHOT_REF)
+      .put(MetadataUpdate.SetProperties.class, SET_PROPERTIES)
+      .put(MetadataUpdate.RemoveProperties.class, REMOVE_PROPERTIES)
+      .put(MetadataUpdate.SetLocation.class, SET_LOCATION)
+      .build();
+
+  public static String toJson(MetadataUpdate metadataUpdate) {
+    return toJson(metadataUpdate, false);
+  }
+
+  public static String toJson(MetadataUpdate metadataUpdate, boolean pretty) {
+    try {
+      StringWriter writer = new StringWriter();
+      JsonGenerator generator = JsonUtil.factory().createGenerator(writer);
+      if (pretty) {
+        generator.useDefaultPrettyPrinter();
+      }
+      toJson(metadataUpdate, generator);
+      generator.flush();
+      return writer.toString();
+    } catch (IOException e) {
+      throw new UncheckedIOException(String.format("Failed to write metadata update json for: %s", metadataUpdate), e);
+    }
+  }
+
+  public static void toJson(MetadataUpdate metadataUpdate, JsonGenerator generator) throws IOException {
+    String updateAction = ACTIONS.get(metadataUpdate.getClass());
+
+    // Add shared `action` field used to differentiate between implementations of MetadataUpdate
+    generator.writeStartObject();
+    generator.writeStringField(ACTION, updateAction);
+
+    // Write subclass specific fields
+    switch (updateAction) {
+      case ASSIGN_UUID:
+        throw new UnsupportedOperationException("Not Implemented: MetadataUpdate#toJson for AssignUUID");
+      case UPGRADE_FORMAT_VERSION:
+        writeUpgradeFormatVersion(metadataUpdate, generator);
+        break;
+      default:
+        throw new IllegalArgumentException(
+            String.format("Cannot convert metadata update to json. Unrecognized action: %s", updateAction));
+    }
+
+    generator.writeEndObject();
+  }
+
+  /**
+   * Read MetadataUpdate from a JSON string.
+   *
+   * @param json a JSON string of a MetadataUpdate
+   * @return a MetadataUpdate object
+   */
+  public static MetadataUpdate fromJson(String json) {
+    try {
+      return fromJson(JsonUtil.mapper().readValue(json, JsonNode.class));
+    } catch (IOException e) {
+      throw new UncheckedIOException("Failed to read JSON string: " + json, e);
+    }
+  }
+
+  public static MetadataUpdate fromJson(JsonNode jsonNode) {
+    Preconditions.checkArgument(jsonNode != null && jsonNode.isObject(),
+        "Cannot parse metadata update from non-object value: %s", jsonNode);
+    Preconditions.checkArgument(jsonNode.hasNonNull(ACTION), "Cannot parse metadata update. Missing field: action");
+    String action = JsonUtil.getString(ACTION, jsonNode).toLowerCase(Locale.ROOT);
+
+    switch (action) {
+      case ASSIGN_UUID:
+        throw new UnsupportedOperationException("Not implemented: AssignUUID");
+      case UPGRADE_FORMAT_VERSION:
+        return readAsUpgradeFormatVersion(jsonNode);
```

Comment:
I will add the implementations for the other classes in a follow up PR after verifying that this is acceptable for others.